x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java

```diff
@@ -1405,6 +1426,326 @@ public void testSecurityIndexStateChangeWillInvalidateApiKeyCaches() throws Exce
+            .filter(s -> s.service().getDocCache().get(apiKey2.v1()) != null)
+            .findFirst()
+            .orElseThrow();
+        final var serviceForDoc1 = serviceWithNameForDoc1.service();
+        final var serviceForDoc2 = serviceWithNameForDoc2.service();
+        assertNotNull(serviceForDoc1.getFromCache(apiKey1.v1()));
+        assertNotNull(serviceForDoc2.getFromCache(apiKey2.v1()));
+
+        final boolean sameServiceNode = serviceWithNameForDoc1 == serviceWithNameForDoc2;
+        if (sameServiceNode) {
+            assertEquals(2, serviceForDoc1.getDocCache().count());
+        } else {
+            assertEquals(1, serviceForDoc1.getDocCache().count());
+            assertEquals(1, serviceForDoc2.getDocCache().count());
+        }
+
+        final int serviceForDoc1AuthCacheCount = serviceForDoc1.getApiKeyAuthCache().count();
+        final int serviceForDoc2AuthCacheCount = serviceForDoc2.getApiKeyAuthCache().count();
+
+        // Update the first key
+        final PlainActionFuture<UpdateApiKeyResponse> listener = new PlainActionFuture<>();
+        serviceForDoc1.updateApiKey(
+            fileRealmAuth(serviceWithNameForDoc1.nodeName(), ES_TEST_ROOT_USER, ES_TEST_ROOT_ROLE),
+            new UpdateApiKeyRequest(apiKey1.v1(), List.of(), null),
+            Set.of(),
+            listener
+        );
+        final var response = listener.get();
+        assertNotNull(response);
+        assertTrue(response.isUpdated());
+
+        // The cache entry should be gone for the first key
+        if (sameServiceNode) {
+            assertEquals(1, serviceForDoc1.getDocCache().count());
+            assertNull(serviceForDoc1.getDocCache().get(apiKey1.v1()));
+            assertNotNull(serviceForDoc1.getDocCache().get(apiKey2.v1()));
+        } else {
+            assertEquals(0, serviceForDoc1.getDocCache().count());
+            assertEquals(1, serviceForDoc2.getDocCache().count());
+        }
+
+        // Auth cache has not been affected
+        assertEquals(serviceForDoc1AuthCacheCount, serviceForDoc1.getApiKeyAuthCache().count());
+        assertEquals(serviceForDoc2AuthCacheCount, serviceForDoc2.getApiKeyAuthCache().count());
+    }
+
+    private static Authentication fileRealmAuth(String nodeName, String userName, String roleName) {
+        boolean includeDomain = randomBoolean();
+        final var realmName = "file";
+        final var realmType = FileRealmSettings.TYPE;
+        return Authentication.newRealmAuthentication(
+            new User(userName, roleName),
+            new Authentication.RealmRef(
+                realmName,
+                realmType,
+                nodeName,
+                includeDomain
+                    ? new RealmDomain(
+                        ESTestCase.randomAlphaOfLengthBetween(3, 8),
+                        Set.of(new RealmConfig.RealmIdentifier(realmType, realmName))
+                    )
+                    : null
+            )
+        );
+    }
```

Comment:
Besides the realm authentication for the user itself, the update should also be possible for:
(1) A token of the user 
(2) Another user run as this user
(3) This username but from a different realm of the same domain

So I think this can use a bit more randomness. Using something like
```
AuthenticationTestHelper.builder().user(...).realmRef(...).build()
```
can address (1) and (2). For (3), we need to hand roll a random domain since the test helper does not provide one that is helpful here (or maybe we can add one test helper method to do that).

Reply:
Since my plan is to update these tests to use the client and transport actions directly I think some of this might actually become redundant/would need to be tweaked since we will need "real" users. Would it be fine to address 1 and 2 here, because it's straight-forward but keep 3 for the next PR?

Reply:
Sounds good to me.